client/signup/hooks/use-branch-steps.ts

```diff
@@ -1,32 +1,33 @@
 import React from 'react';
-import { useDispatch } from 'react-redux';
+import { useSelector, useDispatch } from 'react-redux';
 import flows from 'calypso/signup/config/flows';
-import { addExcludeSteps, removeExcludeSteps } from 'calypso/state/signup/flow/actions';
+import { addExcludedSteps, removeExcludedSteps } from 'calypso/state/signup/flow/actions';
+import { getSignupProgress } from 'calypso/state/signup/progress/selectors';
+import type { Dependencies } from 'calypso/signup/types';
 
-type BranchSteps = ( excludeSteps: string[] ) => void;
+type GetExcludedSteps = ( providedDependencies?: Dependencies ) => string[];
 
-const memoExcludeSteps: { [ key: string ]: string[] } = {};
+type BranchSteps = ( providedDependencies: Dependencies ) => void;
 
 /**
  * This hook returns the function to do branch steps. Ensure this function is called before submitSignupStep
  * because we process all of the steps immediately after submitting.
  * Also, it will clean up the exclude steps when the component mounts as the user might go back a step
  */
-const useBranchSteps = ( stepName: string ): BranchSteps => {
+const useBranchSteps = ( stepName: string, getExcludedSteps: GetExcludedSteps ): BranchSteps => {
 	const dispatch = useDispatch();
-	const branchSteps = ( excludeSteps: string[] ) => {
-		flows.excludeSteps( excludeSteps );
-		memoExcludeSteps[ stepName ] = excludeSteps;
-		dispatch( addExcludeSteps( excludeSteps ) );
+	const signupProgress = useSelector( getSignupProgress );
+	const branchSteps = ( providedDependencies: Dependencies ) => {
+		const excludedSteps = getExcludedSteps( providedDependencies ) || [];
+		flows.excludeSteps( excludedSteps );
```

Comment:
This is more of a question rather a review:

Why do we need to store the excluded steps in two places? (One in `Flow` and one in the redux store)

Can we somehow consolidate them?

Reply:
As the `excludedSteps` in `Flow` is only stored in memory, we should have a way to restore the excluded steps after coming back from 3rd party payment to ensure the Hero Flow works well. So, I make a change to persist `excludedSteps` if those steps are excluded by `useBranchSteps` hook. You can see there are some other places using `excludedSteps` and I think we should not persist them. I know it's not a good idea to keep it in two places but I don't want to make a big change now as `useBranchSteps` is a short-term solution (See pdgK6S-bZ-p2) and it might be removed after we migrate to gutenboarding tech structures (See https://github.com/Automattic/wp-calypso/pull/58560).

Reply:
I see, we want to persist the excluded steps to the local storage.

OK, this makes more sense now. Thanks for linking to the contexts as well!

Reply:
It is unfortunate to have this state in two places, but I agree with the rationale 😮‍💨